pkg/typesystem/connected_types.go

```diff
@@ -0,0 +1,142 @@
+
+		for subjectType, terminalRelations := range cache {
+			for _, terminalRelation := range terminalRelations {
+				terminalTypesAndRelations = append(terminalTypesAndRelations, terminalTypesAndRelation{
+					terminalTypes:    []string{subjectType},
+					terminalRelation: terminalRelation,
+				})
+			}
+		}
+
+		return terminalTypesAndRelations
+	}
+
+	relation, ok := t.relations[typeName][relationName]
+	if !ok {
+		return []terminalTypesAndRelation{}
+	}
+
+	switch rw := rewrite.GetUserset().(type) {
+	case *openfgav1.Userset_This:
+		assignableTypes := []string{}
+		var terminalRelation string
+
+		for _, assignableType := range relation.GetTypeInfo().GetDirectlyRelatedUserTypes() {
+			assignableTypes = append(assignableTypes, assignableType.GetType())
+			terminalRelation = relationName
+		}
+
+		return []terminalTypesAndRelation{
+			{
+				terminalTypes:    assignableTypes,
+				terminalRelation: terminalRelation,
+			},
+		}
+	case *openfgav1.Userset_ComputedUserset:
+		computedRelationName := rw.ComputedUserset.GetRelation()
+		computedRelation := t.relations[typeName][computedRelationName]
+		return t.getTerminalUserTypesAndRelation(typeName, computedRelationName, computedRelation.GetRewrite())
+	case *openfgav1.Userset_TupleToUserset:
+		tuplesetRelationName := rw.TupleToUserset.GetTupleset().GetRelation()
+		computedRelationName := rw.TupleToUserset.GetComputedUserset().GetRelation()
+
+		tuplesetRelation, ok := t.relations[typeName][tuplesetRelationName]
+		if !ok {
+			return []terminalTypesAndRelation{}
+		}
+
+		for _, assignableType := range tuplesetRelation.GetTypeInfo().GetDirectlyRelatedUserTypes() {
+			assignableTypeName := assignableType.GetType()
+
+			if assignableRelation, ok := t.relations[assignableTypeName][computedRelationName]; ok {
+				return t.getTerminalUserTypesAndRelation(assignableTypeName, computedRelationName, assignableRelation.GetRewrite())
+			}
+		}
+
+	case *openfgav1.Userset_Union:
+		assignableTypesForRelation := []terminalTypesAndRelation{}
+
+		for _, child := range rw.Union.GetChild() {
+			tt := t.getTerminalUserTypesAndRelation(typeName, relationName, child)
+			assignableTypesForRelation = append(assignableTypesForRelation, tt...)
+		}
+
+		return assignableTypesForRelation
+
+	}
+	// case *openfgav1.Userset_Intersection:
+	// 	intersectionAssignableTypesAndRelations := []terminalTypesAndRelation{}
+
+	// 	for _, child := range rw.Intersection.GetChild() {
+	// 		_ = t.getTerminalUserTypesAndRelation(typeName, relationName, child)
+	// 	}
+
+	// 	return intersectionAssignableTypesAndRelations
+	// }
+
+	// case *openfgav1.Userset_Difference:
+	// 	// All the children must have an entrypoint.
+	// 	baseTypes, baseTerminalRelation := t.getTerminalUserTypesAndRelation(typeName, relationName, rw.Difference.GetBase())
+	// 	subTypes, subTerminalRelation := t.getTerminalUserTypesAndRelation(typeName, relationName, rw.Difference.GetSubtract())
+
+	// 	return
+	// }
```

Comment:
Intersection and difference should be basically the same as union: gather all the children and then append all the types you find

Reply:
It would be incorrect to treat union and intersection the same because intersection requires an overlap of common types.

Reply:
Is that required for this work? It sounds like the fix for https://github.com/openfga/openfga/issues/1595